collector/lib/ResourceSelector.cpp

```diff
@@ -0,0 +1,96 @@
+//
+// Created by Robby Cochran on 3/28/24.
+//
+
+#include "ResourceSelector.h"
+
+#include <regex>
+
+#include "Hash.h"
+
+namespace collector {
+
+bool ResourceSelector::IsRuleValueFollowed(const storage::RuleValue& value, const std::string& resource_name) {
+  if (value.match_type() == storage::REGEX) {
+    std::regex pattern(value.value());
+    return std::regex_match(resource_name, pattern);
+  } else {
+    return (resource_name == value.value());
+  }
+}
+
+bool ResourceSelector::IsRuleFollowed(const storage::SelectorRule& rule, const std::string& resource_name) {
+  if (rule.operator_() == storage::BooleanOperator::OR) {
+    for (const auto& value : rule.values()) {
+      if (IsRuleValueFollowed(value, resource_name)) {
+        return true;
+      }
+    }
+  } else if (rule.operator_() == storage::BooleanOperator::AND) {
+    for (const auto& value : rule.values()) {
+      if (!IsRuleValueFollowed(value, resource_name)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  return false;
+}
+
+bool ResourceSelector::IsResourceInResourceSelector(const storage::ResourceSelector& rs, const std::string& resource_type, const std::string& resource_name) {
+  for (const auto& rule : rs.rules()) {
+    if (rule.field_name() == resource_type && IsRuleFollowed(rule, resource_name)) {
+      return true;
+    }
+  }
+
+  return false;
+}
```

Comment:
This code was incorrect. There can be multiple SelectorRules in one ResourceSelector with the same field_name, i.e namespace or cluster. In such cases all of the SelectorRules need to evaluate to true. The code above is incorrect, because it would return true if the first SelectorRule is true and the second is false. After fixing the code using `all_of` produced some ugly code. I did include it, but I commented it out.

Reply:
I don't know what I was thinking here. `all_of` works well with the fix.